packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,981 @@
+      // owner of the child with the same id will have given up that child by
+      // then.
+      _childrenToAdd.putIfAbsent(child.id, () => <RestorationBucket>[]).add(child);
+      _markNeedsSerialization();
+      return;
+    }
+    _finalizeAddChildData(child);
+    _markNeedsSerialization();
+  }
+
+  void _finalizeAddChildData(RestorationBucket child) {
+    assert(_claimedChildren[child.id] == null);
+    assert(_rawChildren[child.id.value] == null);
+    _claimedChildren[child.id] = child;
+    _rawChildren[child.id.value] = child._rawData;
+  }
+
+  void _visitChildren(_BucketVisitor visitor, {bool concurrentModification = false}) {
+    Iterable<RestorationBucket> children = _claimedChildren.values
+        .followedBy(_childrenToAdd.values.expand((List<RestorationBucket> buckets) => buckets));
+    if (concurrentModification) {
+      children = children.toList(growable: false);
+    }
+    children.forEach(visitor);
+  }
+
+  // Bucket management
+
+  /// Changes the [RestorationId] under which the bucket is stored in its parent
+  /// to `newId`.
+  ///
+  /// No-op if the bucket is already stored under the provided id.
+  ///
+  /// If another owner has already claimed a bucket with the provided `newId` an
+  /// exception will be thrown at the end of the current frame unless the other
+  /// owner has deleted its bucket by calling [dispose], [rename]ed it using
+  /// another ID, or has moved it to a new parent via [adoptChild].
+  void rename(RestorationId newId) {
+    assert(_debugAssertNotDisposed());
+    assert(newId != null);
+    assert(_parent != null);
+    if (newId == id) {
+      return;
+    }
+    _parent._removeChildData(this);
+    _id = newId;
+    _parent._addChildData(this);
+  }
+
+  /// Deletes the bucket and all the data stored in it from the bucket
+  /// hierarchy.
+  ///
+  /// After [dispose] has been called, the data stored in this bucket and its
+  /// children are no longer part of the app's restoration data. The data
+  /// originally stored in the bucket will not be available again when the
+  /// application is restored to this state in the future. It is up to the
+  /// owners of the children to either move them (via [adoptChild]) to a new
+  /// parent that is still part of the bucket hierarchy or to [dispose] of them
+  /// as well.
+  ///
+  /// This method must only be called by the object's owner.
+  @override
+  void dispose() {
+    assert(_debugAssertNotDisposed());
+    _debugDisposed = true;
+    _parent?._removeChildData(this);
+    _parent = null;
+    _updateManager(null);
+    _visitChildren(_dropChild, concurrentModification: true);
+    _claimedChildren.clear();
+    _childrenToAdd.clear();
+    super.dispose();
+  }
+
+  @override
+  String toString() => '${objectRuntimeType(this, 'RestorationBucket')}(id: $id, owner: $debugOwner)';
+
+  bool _debugDisposed = false;
+  bool _debugAssertNotDisposed() {
+    assert(() {
+      if (_debugDisposed == null) {
```

Comment:
Shouldn't this be `if (_debugDisposed)`?  When would it be null?

Reply:
You're right, fixed it. (I think this was a c&p error)